server/driver_tidb.go

```diff
@@ -149,9 +150,22 @@ func (ts *TiDBStatement) Reset() {
 // Close implements PreparedStatement Close method.
 func (ts *TiDBStatement) Close() error {
 	//TODO close at tidb level
-	err := ts.ctx.DropPreparedStmt(ts.id)
-	if err != nil {
-		return err
+	if ts.ctx.GetSessionVars().TxnCtx != nil && ts.ctx.GetSessionVars().TxnCtx.CouldRetry {
+		err := ts.ctx.DropPreparedStmt(ts.id)
+		if err != nil {
+			return err
+		}
+	} else {
+		if core.PreparedPlanCacheEnabled() {
+			preparedPointer := ts.ctx.GetSessionVars().PreparedStmts[ts.id]
+			preparedObj, ok := preparedPointer.(*core.CachedPrepareStmt)
+			if !ok {
+				return errors.Errorf("invalid CachedPrepareStmt type")
```

Comment:
Could you use a pre-defined error for it?

Reply:
There are 5 other usages, I think it's ok to keep it. To change all of them, it would be better to do it in another PR.

Reply:
Ok.